File: org.kevoree.microsandbox.watchdog/scratch/LangWalker.java
// $ANTLR 2.7.2: "treewalk.g" -> "LangWalker.java"$

import antlr.TreeParser;
import antlr.Token;
import antlr.collections.AST;
import antlr.RecognitionException;
import antlr.ANTLRException;
import antlr.NoViableAltException;
import antlr.MismatchedTokenException;
import antlr.SemanticException;
import antlr.collections.impl.BitSet;
import antlr.ASTPair;
import antlr.collections.impl.ASTArray;


public class LangWalker extends antlr.TreeParser       implements LangParserTokenTypes
 {
public LangWalker() {
	tokenNames = _tokenNames;
}

	public final void block(AST _t) throws RecognitionException {
		
		AST block_AST_in = (AST)_t;
		
		try {      // for error handling
			AST __t20154 = _t;
			AST tmp14_AST_in = (AST)_t;
			match(_t,BLOCK);
			_t = _t.getFirstChild();
			{
			int _cnt20156=0;
			_loop20156:
			do {
				if (_t==null) _t=ASTNULL;
				if ((_tokenSet_0.member(_t.getType()))) {
					stat(_t);
					_t = _retTree;
				}
				else {
					if ( _cnt20156>=1 ) { break _loop20156; } else {throw new NoViableAltException(_t);}
				}
				
				_cnt20156++;
			} while (true);
			}
			_t = __t20154;
			_t = _t.getNextSibling();
		}
		catch (RecognitionException ex) {
			reportError(ex);
			if (_t!=null) {_t = _t.getNextSibling();}
		}
		_retTree = _t;
	}
	
	public final void stat(AST _t) throws RecognitionException {
		
		AST stat_AST_in = (AST)_t;
		
		try {      // for error handling
			if (_t==null) _t=ASTNULL;
			switch ( _t.getType()) {
			case TK_if:
			{
				AST __t20158 = _t;
				AST tmp15_AST_in = (AST)_t;
				match(_t,TK_if);
				_t = _t.getFirstChild();
				expr(_t);
				_t = _retTree;
				stat(_t);
				_t = _retTree;
				{
				if (_t==null) _t=ASTNULL;
				switch ( _t.getType()) {
				case BLOCK:
				case ASSIGN:
				case TK_if:
				case TK_while:
				case EXPR:
				{
					stat(_t);
					_t = _retTree;
					break;
				}
				case 3:
				{
					break;
				}
				default:
				{
					throw new NoViableAltException(_t);
				}
				}
				}
				_t = __t20158;
				_t = _t.getNextSibling();
				break;
			}
			case TK_while:
			{
				AST __t20160 = _t;
				AST tmp16_AST_in = (AST)_t;
				match(_t,TK_while);
				_t = _t.getFirstChild();
				expr(_t);
				_t = _retTree;
				stat(_t);
				_t = _retTree;
				_t = __t20160;
				_t = _t.getNextSibling();
				break;
			}
			case EXPR:
			{
				expr(_t);
				_t = _retTree;
				break;
			}
			case BLOCK:
			{
				block(_t);
				_t = _retTree;
				break;
			}
			case ASSIGN:
			{
				AST __t20161 = _t;
				AST tmp17_AST_in = (AST)_t;
				match(_t,ASSIGN);
				_t = _t.getFirstChild();
				AST tmp18_AST_in = (AST)_t;
				match(_t,ID);
				_t = _t.getNextSibling();
				expr(_t);
				_t = _retTree;
				_t = __t20161;
				_t = _t.getNextSibling();
				break;
			}
			default:
			{
				throw new NoViableAltException(_t);
			}
			}
		}
		catch (RecognitionException ex) {
			reportError(ex);
			if (_t!=null) {_t = _t.getNextSibling();}
		}
		_retTree = _t;
	}
	
	public final void expr(AST _t) throws RecognitionException {
		
		AST expr_AST_in = (AST)_t;
		AST a = null;
		AST b = null;
		
		try {      // for error handling
			AST __t20163 = _t;
			AST tmp19_AST_in = (AST)_t;
			match(_t,EXPR);
			_t = _t.getFirstChild();
			{
			if (_t==null) _t=ASTNULL;
			switch ( _t.getType()) {
			case ID:
			{
				a = (AST)_t;
				match(_t,ID);
				_t = _t.getNextSibling();
				System.out.println("found ID "+a.getText());
				break;
			}
			case INT:
			{
				b = (AST)_t;
				match(_t,INT);
				_t = _t.getNextSibling();
				System.out.println("found INT "+b.getText());
				break;
			}
			default:
			{
				throw new NoViableAltException(_t);
			}
			}
			}
			_t = __t20163;
			_t = _t.getNextSibling();
		}
		catch (RecognitionException ex) {
			reportError(ex);
			if (_t!=null) {_t = _t.getNextSibling();}
		}
		_retTree = _t;
	}
	
	
	public static final String[] _tokenNames = {
		"<0>",
		"EOF",
		"<2>",
		"NULL_TREE_LOOKAHEAD",
		"BLOCK",
		"LCURLY",
		"RCURLY",
		"ID",
		"ASSIGN",
		"SEMI",
		"\"if\"",
		"LPAREN",
		"RPAREN",
		"\"else\"",
		"\"while\"",
		"INT",
		"EXPR",
		"WS",
		"DIGIT"
	};
	
	private static final long[] mk_tokenSet_0() {
		long[] data = { 83216L, 0L};
		return data;
	}
	public static final BitSet _tokenSet_0 = new BitSet(mk_tokenSet_0());
	}
	

File: org.kevoree.microsandbox.watchdog/scratch/LangWalker.cpp
/* $ANTLR 2.7.2: "treewalk.g" -> "LangWalker.cpp"$ */
#include "LangWalker.hpp"
#include <antlr/Token.hpp>
#include <antlr/AST.hpp>
#include <antlr/NoViableAltException.hpp>
#include <antlr/MismatchedTokenException.hpp>
#include <antlr/SemanticException.hpp>
#include <antlr/BitSet.hpp>
#line 1 "treewalk.g"
#line 11 "LangWalker.cpp"
LangWalker::LangWalker()
	: ANTLR_USE_NAMESPACE(antlr)TreeParser() {
}

void LangWalker::block(ANTLR_USE_NAMESPACE(antlr)RefAST _t) {
	ANTLR_USE_NAMESPACE(antlr)RefAST block_AST_in = _t;
	
	try {      // for error handling
		ANTLR_USE_NAMESPACE(antlr)RefAST __t21560 = _t;
		ANTLR_USE_NAMESPACE(antlr)RefAST tmp14_AST_in = _t;
		match(_t,BLOCK);
		_t = _t->getFirstChild();
		{ // ( ... )+
		int _cnt21562=0;
		for (;;) {
			if (_t == ANTLR_USE_NAMESPACE(antlr)nullAST )
				_t = ASTNULL;
			if ((_tokenSet_0.member(_t->getType()))) {
				stat(_t);
				_t = _retTree;
			}
			else {
				if ( _cnt21562>=1 ) { goto _loop21562; } else {throw ANTLR_USE_NAMESPACE(antlr)NoViableAltException(_t);}
			}
			
			_cnt21562++;
		}
		_loop21562:;
		}  // ( ... )+
		_t = __t21560;
		_t = _t->getNextSibling();
	}
	catch (ANTLR_USE_NAMESPACE(antlr)RecognitionException& ex) {
		reportError(ex);
		if ( _t != ANTLR_USE_NAMESPACE(antlr)nullAST )
			_t = _t->getNextSibling();
	}
	_retTree = _t;
}

void LangWalker::stat(ANTLR_USE_NAMESPACE(antlr)RefAST _t) {
	ANTLR_USE_NAMESPACE(antlr)RefAST stat_AST_in = _t;
	
	try {      // for error handling
		if (_t == ANTLR_USE_NAMESPACE(antlr)nullAST )
			_t = ASTNULL;
		switch ( _t->getType()) {
		case TK_if:
		{
			ANTLR_USE_NAMESPACE(antlr)RefAST __t21564 = _t;
			ANTLR_USE_NAMESPACE(antlr)RefAST tmp15_AST_in = _t;
			match(_t,TK_if);
			_t = _t->getFirstChild();
			expr(_t);
			_t = _retTree;
			stat(_t);
			_t = _retTree;
			{
			if (_t == ANTLR_USE_NAMESPACE(antlr)nullAST )
				_t = ASTNULL;
			switch ( _t->getType()) {
			case BLOCK:
			case ASSIGN:
			case TK_if:
			case TK_while:
			case EXPR:
			{
				stat(_t);
				_t = _retTree;
				break;
			}
			case 3:
			{
				break;
			}
			default:
			{
				throw ANTLR_USE_NAMESPACE(antlr)NoViableAltException(_t);
			}
			}
			}
			_t = __t21564;
			_t = _t->getNextSibling();
			break;
		}
		case TK_while:
		{
			ANTLR_USE_NAMESPACE(antlr)RefAST __t21566 = _t;
			ANTLR_USE_NAMESPACE(antlr)RefAST tmp16_AST_in = _t;
			match(_t,TK_while);
			_t = _t->getFirstChild();
			expr(_t);
			_t = _retTree;
			stat(_t);
			_t = _retTree;
			_t = __t21566;
			_t = _t->getNextSibling();
			break;
		}
		case EXPR:
		{
			expr(_t);
			_t = _retTree;
			break;
		}
		case BLOCK:
		{
			block(_t);
			_t = _retTree;
			break;
		}
		case ASSIGN:
		{
			ANTLR_USE_NAMESPACE(antlr)RefAST __t21567 = _t;
			ANTLR_USE_NAMESPACE(antlr)RefAST tmp17_AST_in = _t;
			match(_t,ASSIGN);
			_t = _t->getFirstChild();
			ANTLR_USE_NAMESPACE(antlr)RefAST tmp18_AST_in = _t;
			match(_t,ID);
			_t = _t->getNextSibling();
			expr(_t);
			_t = _retTree;
			_t = __t21567;
			_t = _t->getNextSibling();
			break;
		}
		default:
		{
			throw ANTLR_USE_NAMESPACE(antlr)NoViableAltException(_t);
		}
		}
	}
	catch (ANTLR_USE_NAMESPACE(antlr)RecognitionException& ex) {
		reportError(ex);
		if ( _t != ANTLR_USE_NAMESPACE(antlr)nullAST )
			_t = _t->getNextSibling();
	}
	_retTree = _t;
}

void LangWalker::expr(ANTLR_USE_NAMESPACE(antlr)RefAST _t) {
	ANTLR_USE_NAMESPACE(antlr)RefAST expr_AST_in = _t;
	ANTLR_USE_NAMESPACE(antlr)RefAST a = ANTLR_USE_NAMESPACE(antlr)nullAST;
	ANTLR_USE_NAMESPACE(antlr)RefAST b = ANTLR_USE_NAMESPACE(antlr)nullAST;
	
	try {      // for error handling
		ANTLR_USE_NAMESPACE(antlr)RefAST __t21569 = _t;
		ANTLR_USE_NAMESPACE(antlr)RefAST tmp19_AST_in = _t;
		match(_t,EXPR);
		_t = _t->getFirstChild();
		{
		if (_t == ANTLR_USE_NAMESPACE(antlr)nullAST )
			_t = ASTNULL;
		switch ( _t->getType()) {
		case ID:
		{
			a = _t;
			match(_t,ID);
			_t = _t->getNextSibling();
#line 54 "treewalk.g"
			std::cout << "found ID "<<a->getText() << std::endl;
#line 173 "LangWalker.cpp"
			break;
		}
		case INT:
		{
			b = _t;
			match(_t,INT);
			_t = _t->getNextSibling();
#line 55 "treewalk.g"
			std::cout << "found INT "<<b->getText() << std::endl;
#line 183 "LangWalker.cpp"
			break;
		}
		default:
		{
			throw ANTLR_USE_NAMESPACE(antlr)NoViableAltException(_t);
		}
		}
		}
		_t = __t21569;
		_t = _t->getNextSibling();
	}
	catch (ANTLR_USE_NAMESPACE(antlr)RecognitionException& ex) {
		reportError(ex);
		if ( _t != ANTLR_USE_NAMESPACE(antlr)nullAST )
			_t = _t->getNextSibling();
	}
	_retTree = _t;
}

void LangWalker::initializeASTFactory( ANTLR_USE_NAMESPACE(antlr)ASTFactory& factory )
{
}
const char* LangWalker::tokenNames[] = {
	"<0>",
	"EOF",
	"<2>",
	"NULL_TREE_LOOKAHEAD",
	"BLOCK",
	"LCURLY",
	"RCURLY",
	"ID",
	"ASSIGN",
	"SEMI",
	"\"if\"",
	"LPAREN",
	"RPAREN",
	"\"else\"",
	"\"while\"",
	"INT",
	"EXPR",
	"WS_",
	"DIGIT",
	0
};

const unsigned long LangWalker::_tokenSet_0_data_[] = { 83216UL, 0UL, 0UL, 0UL };
// BLOCK ASSIGN "if" "while" EXPR 
const ANTLR_USE_NAMESPACE(antlr)BitSet LangWalker::_tokenSet_0(_tokenSet_0_data_,4);


